File: frontend/src/api/brain.ts
import { API_BASE, apiFetch } from "./client";

export interface BrainStats {
  enabled: boolean;
  entities: number;
  facts: number;
  episodes: number;
  preferences: number;
  chunks: number;
  documents: number;
}

export interface BrainGraphNode {
  id: string;
  name: string;
  entity_type: string | null;
  confidence: number;
}

export interface BrainGraphEdge {
  source: string;
  type: string;
  target: string;
  weight: number;
}

export interface BrainGraph {
  nodes: BrainGraphNode[];
  edges: BrainGraphEdge[];
}

export interface BrainEntity {
  entity_id: string;
  name: string;
  entity_type: string | null;
  aliases: string[];
  description: string | null;
  confidence: number;
  created_at: string | null;
  last_accessed_at: string | null;
}

export interface BrainEpisode {
  episode_id: string;
  kind: string;
  summary: string;
  occurred_at: string | null;
  conversation_id: number | null;
  participants: string[];
}

export interface BrainPreference {
  pref_id: string;
  key: string;
  value: string;
  confidence: number;
  valid_from: string | null;
}

export interface BrainFact {
  fact_id: string;
  subject_id: string;
  predicate: string;
  object_text: string;
  valid_from: string | null;
  valid_to: string | null;
  confidence: number;
  source_doc_id: string | null;
}

export type BrainMemoryType = "episode" | "fact" | "preference";

export interface BrainFactCorrection {
  object_text: string;
  predicate?: string | null;
  confidence?: number | null;
}

export interface BrainPreferenceUpdate {
  value: string;
  confidence?: number | null;
}

export async function getBrainStats(): Promise<BrainStats> {
  const res = await apiFetch(`${API_BASE}/brain/stats`);
  if (!res.ok) throw new Error("加载大脑统计失败");
  return res.json();
}

export async function getBrainGraph(limit = 80): Promise<BrainGraph> {
  const res = await apiFetch(`${API_BASE}/brain/graph?limit=${limit}`);
  if (!res.ok) throw new Error("加载知识图谱失败");
  return res.json();
}

export async function listBrainEntities(limit = 200, offset = 0): Promise<BrainEntity[]> {
  const res = await apiFetch(`${API_BASE}/brain/entities?limit=${limit}&offset=${offset}`);
  if (!res.ok) throw new Error("加载实体失败");
  return res.json();
}

export async function listBrainEpisodes(limit = 100, offset = 0): Promise<BrainEpisode[]> {
  const res = await apiFetch(`${API_BASE}/brain/episodes?limit=${limit}&offset=${offset}`);
  if (!res.ok) throw new Error("加载记忆失败");
  return res.json();
}

export async function listBrainPreferences(): Promise<BrainPreference[]> {
  const res = await apiFetch(`${API_BASE}/brain/preferences`);
  if (!res.ok) throw new Error("加载偏好失败");
  return res.json();
}

export async function listBrainFacts(entityId?: string, limit = 200): Promise<BrainFact[]> {
  const qs = entityId ? `&entity_id=${encodeURIComponent(entityId)}` : "";
  const res = await apiFetch(`${API_BASE}/brain/facts?limit=${limit}${qs}`);
  if (!res.ok) throw new Error("加载事实失败");
  return res.json();
}

export async function deleteBrainMemory(memoryType: BrainMemoryType, memoryId: string): Promise<void> {
  const res = await apiFetch(`${API_BASE}/brain/memories/${memoryType}/${encodeURIComponent(memoryId)}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error("删除记忆失败");
}

export async function correctBrainFact(
  factId: string,
  payload: BrainFactCorrection,
): Promise<BrainFact> {
  const res = await apiFetch(`${API_BASE}/brain/facts/${encodeURIComponent(factId)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error("纠正事实失败");
  return res.json();
}

export async function updateBrainPreference(
  prefId: string,
  payload: BrainPreferenceUpdate,
): Promise<BrainPreference> {
  const res = await apiFetch(`${API_BASE}/brain/preferences/${encodeURIComponent(prefId)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error("更新偏好失败");
  return res.json();
}

File: backend/src/schemas/brain.py
"""大脑只读 API 响应模型：知识图谱 / 记忆 / 偏好 / 统计。"""

from pydantic import BaseModel, Field


class BrainStats(BaseModel):
    enabled: bool
    entities: int = 0
    facts: int = 0
    episodes: int = 0
    preferences: int = 0
    chunks: int = 0
    documents: int = 0


class BrainEntity(BaseModel):
    entity_id: str
    name: str
    entity_type: str | None = None
    aliases: list[str] = []
    description: str | None = None
    confidence: float = 1.0
    created_at: str | None = None
    last_accessed_at: str | None = None


class BrainEpisode(BaseModel):
    episode_id: str
    kind: str
    summary: str
    occurred_at: str | None = None
    conversation_id: int | None = None
    participants: list[str] = []


class BrainPreference(BaseModel):
    pref_id: str
    key: str
    value: str
    confidence: float = 1.0
    valid_from: str | None = None


class BrainFact(BaseModel):
    fact_id: str
    subject_id: str
    predicate: str
    object_text: str
    valid_from: str | None = None
    valid_to: str | None = None
    confidence: float = 1.0
    source_doc_id: str | None = None


class BrainGraphNode(BaseModel):
    id: str
    name: str
    entity_type: str | None = None
    confidence: float = 1.0


class BrainGraphEdge(BaseModel):
    source: str
    type: str
    target: str
    weight: float = 1.0


class BrainGraph(BaseModel):
    nodes: list[BrainGraphNode]
    edges: list[BrainGraphEdge]


class BrainFactCorrection(BaseModel):
    object_text: str = Field(min_length=1, max_length=2000)
    predicate: str | None = Field(default=None, min_length=1, max_length=200)
    confidence: float | None = Field(default=None, ge=0, le=1)


class BrainPreferenceUpdate(BaseModel):
    value: str = Field(min_length=1, max_length=2000)
    confidence: float | None = Field(default=None, ge=0, le=1)
